test/spec/features/modeling/behavior/RootElementReferenceBehaviorSpec.js

```diff
@@ -316,6 +510,163 @@ describe('features/modeling - root element reference behavior', function() {
           expect(rootElementsOfType).to.have.lengthOf(rootElementsOfTypeCount);
         }));
       });
+
+
+      forEach([
+        'ReceiveTask_noRef',
+        'SendTask_noRef'
+      ], function(type) {
+
+        var id = type;
+
+        var task,
+            rootElement;
+
+        describe('should add on modeling#updateProperties', function() {
+
+
+          beforeEach(inject(function(bpmnFactory, elementRegistry, modeling) {
+
+            // given
+            task = elementRegistry.get(id);
+
+            rootElement = bpmnFactory.create('bpmn:Message', { id: 'NewMessage' });
+
+            // when
+            modeling.updateProperties(task, {
+              messageRef: rootElement,
+            });
+          }));
+
+
+          it('<do>', function() {
+
+            // then
+            expect(hasRootElement(rootElement)).to.be.true;
+          });
+
+
+          it('<undo>', inject(function(commandStack) {
+
+            // when
+            commandStack.undo();
+
+            // then
+            expect(hasRootElement(rootElement)).to.be.false;
+          }));
+
+
+          it('<redo>', inject(function(commandStack) {
+
+            // given
+            commandStack.undo();
+
+            // when
+            commandStack.redo();
+
+            // then
+            expect(hasRootElement(rootElement)).to.be.true;
+          }));
+        });
+
+
+        describe('should add on modeling#updateModdleProperties', function() {
+
+
+          beforeEach(inject(function(bpmnFactory, elementRegistry, modeling) {
+
+            // given
+            task = elementRegistry.get(id);
+            var bo = getBusinessObject(task);
+
+            rootElement = bpmnFactory.create('bpmn:Message', { id: 'NewMessage' });
+
+            // when
+            modeling.updateModdleProperties(task, bo, {
+              messageRef: rootElement,
+            });
+          }));
+
+
+          it('<do>', function() {
+
+            // then
+            expect(hasRootElement(rootElement)).to.be.true;
+          });
+
+
+          it('<undo>', inject(function(commandStack) {
+
+            // when
+            commandStack.undo();
+
+            // then
+            expect(hasRootElement(rootElement)).to.be.false;
+          }));
+
+
+          it('<redo>', inject(function(commandStack) {
+
+            // given
+            commandStack.undo();
+
+            // when
+            commandStack.redo();
+
+            // then
+            expect(hasRootElement(rootElement)).to.be.true;
+          }));
+        });
+
+
+        it('should NOT add on modeling#updateProperties if already added', inject(function(
+            bpmnFactory, bpmnjs, elementRegistry, modeling
+        ) {
+
+          // given
+          task = elementRegistry.get(id);
+
+          var message = bpmnFactory.create('bpmn:Message', { id: 'Message_2' });

```

Comment:
I assume `bpmnFactory#create` triggers a hidden side-effect and adds the message?

Reply:
Not really, the message is added as a reference via https://github.com/bpmn-io/bpmn-js/pull/1855/files#diff-eb9635ee33330272486aec3b24a14c913f6d10c8903f3fdc29b3519908adb005R637, and then the tested behavior reacts on `element.updateProperties` and adds the Message to the definitions#rootElements.

Reply:
So the description should read `should add only once`, right?

Reply:
Hmm it should not add root element if it is already present on the diagram. However, the test case is broken as we just duplicate ID instead of the exact reference. I need to think about it.

Reply:
I will update the test cases so that they clearly show what we want to support.